### PR TITLE
Update unit tests for opcodes

### PR DESCRIPTION
**Pull request by redheadedrod**
This should be the latest code as of 6/2

### CODE REVIEW

File: envi/archs/arm/__init__.py [hunk 1]
```diff
@@ -9,6 +9,12 @@
 from envi.archs.arm.disasm import *
 
 class ArmModule(envi.ArchitectureModule):
+    '''following line before function is to help with different architecture
```
Comment (atlas0fd00m) [inline envi/archs/arm/__init__.py]: purely aesthetic, but please place the quotes on their own line:
'''
following...
...
'''

File: envi/archs/arm/__init__.py [hunk 2]
```diff
@@ -50,6 +56,11 @@ class ThumbModule(envi.ArchitectureModule):
     '''
     This architecture module will *not* shift to ARM mode.  Evar.
     '''
+    '''following line before function is to help with different architecture
```
Comment (atlas0fd00m) [inline envi/archs/arm/__init__.py]: ditto

File: envi/archs/arm/const.py [hunk 3]
```diff
@@ -3,6 +3,32 @@
...
 MODE_THUMBEE    = 3
 
+'''
+bitfields to support different versions
+For simplicity will use 0xFFFF to say all.
+Note that thumb and arm are both represented
+thumbee is shown but may never be fully implimented
+since was already depreciated
+'''
+archBitMask = [
```
Comment (atlas0fd00m) [inline envi/archs/arm/const.py]: please make these "constants":

REV_ARMv4 = 0b.....
REV_ARMv5 = ...
REV_ARMv5T = ...

then group them altogether in a dictionary:
ARCH_REVS = { num : name for name,num in globals().items() if name.startswith('REV_') }

an interactive example:
In [1]: REV_FOO = 1

In [2]: ARCH_REVS = { num : name for name,num in globals().items() if name.startswith('REV_') }

In [3]: ARCH_REVS
Out[3]: {1: 'REV_FOO'}


you don't want to be wielding tuples when actually using the thing.  you want to use the constant, and let it be "the thing"  like:

if self.capabilities & ARVv5T:
    dostuff()

File: envi/archs/arm/const.py [hunk 4]
```diff
@@ -3,6 +3,32 @@
...
+#    architecture   bitmask              dec     hex
+    ('THUMB16',     0b0000000000000001), # 1      1
```
Comment (atlas0fd00m) [inline envi/archs/arm/const.py]: i believe thumb16, thumb2, and thumbee are all defined by the ARMv* items below.  for instance, ARMv5T includes THUMB16.  we should be able to clearly define an architecture using the ARMv* only.  for example, ARMv7M will be THUMB2 only.  ARMv7AR will support THUMB2.  if you'd like THUMB2 to be easily available, make it a bitfield combination of all ARMv* architectures which support THUMB2.

REV_ARMv7 = (REV_ARMv7M | REV_ARMv7AR)
REV_THUMB2 = (REV_ARMv6T2 | REV_ARMv7 | REV_ARMv8)

File: envi/archs/arm/disasm.py [hunk 5]
```diff
@@ -314,7 +314,7 @@ def p_misc(opval, va):
             ArmRegOper(Rm, va=va),
             ArmRegOper(Rs, va=va),
         )
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: good catch!

File: envi/archs/arm/disasm.py [hunk 6]
```diff
@@ -218,7 +218,7 @@ def p_dp_imm_shift(opval, va):
 qop_mnem = ('qadd','qsub','qdadd','qdsub')
 smla_mnem = ('smlabb','smlabt','smlatb','smlatt',)
 smlal_mnem = ('smlalbb','smlalbt','smlaltb','smlaltt',)
-smul_mnem = ('smulbb','smulbt','smultb','smultt',)
+smul_mnem = ('smulbb','smultb','smulbt','smultt',)  #bt and tb are backwards. Note xy bits are actually yx
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: another good catch.  since this encoding is an oddball, please go to the parser and change "xy" to "yx" so it's easy to see the difference.

File: envi/archs/arm/disasm.py [hunk 7]
```diff
@@ -1837,7 +1843,7 @@ def getOperAddr(self, op, emu=None):
         # if U==0, subtract
         addval *= pom
 
-        addr = (Rn + addval) & e_bits.u_maxes[self.tsize]
+        addr = Rn + addval
```
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: please put this back as you found it.  i believe you accidentally made this change.

File: envi/tests/test_arch_arm.py [hunk 8]
```diff
@@ -5,58 +5,594 @@
...
-from envi.archs.h8.regs import *
-from envi.archs.h8.const import *
-from envi.archs.h8.parsers import *
+from envi.archs.arm.regs import *
+from envi.archs.arm.const import *
+from envi.archs.arm.disasm import *
+import binascii   # temporarily included for testing purposes only - allows to print out binary values
 
 
-# OPHEX, VA, repr, flags, emutests
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: please put these back in, as they are ARM-specific, and i just added them.

File: envi/tests/test_arch_arm.py [hunk 9]
```diff
@@ -5,58 +5,594 @@
...
-# FIXME: create list of this from IDA flow below - THIS CURRENT DATA IS FOR H8!  NOT ARM/THUMB
+
+
+''' 
+
+  This dictionary will contain all commands supported by ARM to test
+  Fields will contain following information:
+  archBitMask, ophex, va, repr, flags, emutests
 '''
-        ( '8342', 0x4560, 'add.b #42, r3h', IF_B, () ),
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: these, however, are obviously H8 remnants and can be deleted.

File: envi/tests/test_arch_arm.py [hunk 10]
```diff
@@ -271,49 +814,83 @@ def test_envi_arm_operands(self):
...
+            for arch_mask in range(number_of_archs):
+                test_arch = int(pow(2,arch_mask)) #mask for architecture to test
+                if ((not ranAlready) or (not self.armTestOnce)) and ((archz & test_arch & self.armTestVersion) != 0):
+                    ranAlready = True
+                    arm.ThumbModule.archVersion = arm.ArmModule.archVersion = archBitMask[arch_mask][0]
+                    op = vw.arch.archParseOpcode(bytez.decode('hex'), 0, va)
+                    redoprepr = repr(op).replace(' ','').lower()
+                    redgoodop = reprOp.replace(' ','')
+                    if redoprepr != redgoodop:
+                        print  bytez,redgoodop
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: before we're done, this will all want to be seriously cleaned up (comments, prints, etc...)  but for now, this suits our purposes.  i only state this because the PR makes it more official.

File: envi/archs/arm/disasm.py (hunk 6 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Wasn't sure if the swap should have been done here or at the decoding section but seemed like it was easier to do here and fix later if the emulation was backwards.

Date: Fri, 3 Jun 2016 05:52:32 -0700

File: envi/archs/arm/disasm.py (hunk 7 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Yes was unintentional. Fixed on my copy will push today

Date: Fri, 3 Jun 2016 05:54:26 -0700

File: envi/tests/test_arch_arm.py (hunk 8 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Note: These were not removed, they were moved... See for reference:

''' 
  This dictionary will contain all commands supported by ARM to test
  Fields will contain following information:
  archBitMask, ophex, va, repr, flags, emutests
'''
I will capitalize them to match

Date: Fri, 3 Jun 2016 05:55:33 -0700

File: envi/tests/test_arch_arm.py (hunk 10 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Understood.  And expected. I have been incrementally cleaning this up as I go along.  Trying to only keep what makes sense to help with the creation and debugging of new code.

Date: Fri, 3 Jun 2016 06:01:24 -0700

File: envi/tests/test_arch_arm.py (hunk 9 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Only keeping this section for reference. I will cut it down further. I just need the setup and test versions for examples. Will be removed as soon as first emulation test is included.

Date: Fri, 3 Jun 2016 05:55:51 -0700

File: envi/archs/arm/disasm.py (hunk 6 above)
Comment (atlas0fd00m) [inline envi/archs/arm/disasm.py]: here is correct.  i'm telling you to swap the letters in the name of the variable in the decoder.

File: envi/tests/test_arch_arm.py [hunk 11]
```diff
@@ -5,58 +5,594 @@
...
-from envi.archs.h8.const import *
-from envi.archs.h8.parsers import *
+from envi.archs.arm.regs import *
+from envi.archs.arm.const import *
+from envi.archs.arm.disasm import *
+import binascii   # temporarily included for testing purposes only - allows to print out binary values
 
 
-# OPHEX, VA, repr, flags, emutests
-instrs = [
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: these?  ldr r3 doesn't show up in your list at all.

File: envi/archs/arm/disasm.py (hunk 6 above)
Comment (redheadedrod) [inline envi/archs/arm/disasm.py]: Ok will do.

Date: Fri, 3 Jun 2016 06:55:56 -0700

File: envi/tests/test_arch_arm.py (hunk 11 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Not sure I fully understand what you say is missing.

 If you are referring to the list missing items it SHOULD be the full IDA dump converted to be compatible with this format. If there are items missing from the list then they are missing from the original IDA dump. Luckily I saved the helper module in case we needed to do any further IDA dumps. 

I also saved the original dump listing. 
There are no straight LDR lisings in the IDA dump I have. Just modifications of it such as LDRH



Date: Fri, 3 Jun 2016 06:57:37 -0700

File: envi/archs/arm/const.py (hunk 4 above)
Comment (redheadedrod) [inline envi/archs/arm/const.py]: Ok, I more or less added the thumb stuff mostly for testing purposes.
Using in the manner you mention makes it much easier to make modifications to it later.
  
 It also allowed me to test the thumb variants on their own but I suspect that just calling the ARM version with an address of negative value will do the same. 

Since ARMv7M is basically the same as ARMv7AR minus the ARM code, ie only THUMB2 code, should it be included with them?
I will work through this stuff and post an updated variation.

Date: Fri, 3 Jun 2016 05:39:13 -0700

File: envi/archs/arm/const.py [hunk 12]
```diff
@@ -4,30 +4,61 @@
...
+
+#Not sure, did from memory , needs to be confirmed
+REVT_THUMB16 = (REVS_ARMv5 | REVS_ARMv6)
+REVT_THUMB2  = (REVS_ARMv7 | REVS_ARMv8)
+REVT_THUMBEE = (REVS_ARMv7 | REVS_ARMv8)
+
+#Not sure I understand your request but first attempt:
+#using string as opposed to number for key to make future addtions easier but
+#if key field needs to be a number for performance then will change
+ARCH_REVS = {
```
Comment (atlas0fd00m) [inline envi/archs/arm/const.py]: i actually gave you the code to do what i wanted :)  that's why i gave you the example of running it, so you understood what it did.  look again.

File: envi/tests/test_arch_arm.py [hunk 13]
```diff
@@ -19,10 +19,9 @@
 
 
 ''' 
-
   This dictionary will contain all commands supported by ARM to test
   Fields will contain following information:
-  archBitMask, ophex, va, repr, flags, emutests
+  archBitMask, OPHEX, VA, repr, flags, emutests
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: no thank you.  please keep the CAPITALS to a minimum.  just an aesthetic, but something i feel strongly about.

File: envi/archs/arm/__init__.py [hunk 14]
```diff
@@ -9,6 +9,14 @@
 from envi.archs.arm.disasm import *
 
 class ArmModule(envi.ArchitectureModule):
+    '''
+    Following line before function is to help with different architecture
+       versions. These are not planned to be fully implemented until after ARMv8
+       is implimented. Will be important for including ARMv7M and
+       some other variants that fork from normal instruction path.
+    '''
+    archVersion = ('ARMv4')
```
Comment (atlas0fd00m) [inline envi/archs/arm/__init__.py]: this is a problem.  we're targeting ARMv7+

File: envi/tests/test_arch_arm.py [hunk 15]
```diff
@@ -5,68 +5,577 @@
...
-        ('02209ae7', 0xbfb00000, 'ldr r2, [r10, r2]', 0, ()),
-        ('02209ae6', 0xbfb00000, 'ldr r2, [r10], r2', 0, ()),
-        ('02203ae7', 0xbfb00000, 'ldr r2, [r10, -r2]!', 0, ()),
-        ('0220bae7', 0xbfb00000, 'ldr r2, [r10, r2]!', 0, ()),
-        ('22209ae7', 0xbfb00000, 'ldr r2, [r10, r2 lsr #32]', 0, ()),
-        ('08309fe5', 0xbfb00000, 'ldr r3, [#0xbfb00010]', 0, ()),
-        ('08309fe5', 0xbfb00000, 'ldr r3, [#0xbfb00010]', 0, ()),
-        ]
-# FIXME: create list of this from IDA flow below - THIS CURRENT DATA IS FOR H8!  NOT ARM/THUMB
+        (0xffff, '674503e0', 0x4560, 'and r4, r3, r7, ror #10', 0, ()),
```
Comment (atlas0fd00m) [inline envi/tests/test_arch_arm.py]: please replace all the 0xffff magic number madness with REV_ALL
and set REV_ALL in const.py

File: envi/archs/arm/__init__.py (hunk 14 above)
Comment (redheadedrod) [inline envi/archs/arm/__init__.py]: Just needed a default value. It can be easily changed but the structure is not complete yet. 
Once the CONSTANT stuff is complete I will update this and the code that uses it. 
Date: Fri, 3 Jun 2016 11:00:09 -0700

File: envi/tests/test_arch_arm.py (hunk 15 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Done -

Date: Fri, 3 Jun 2016 11:02:48 -0700

File: envi/tests/test_arch_arm.py (hunk 13 above)
Comment (redheadedrod) [inline envi/tests/test_arch_arm.py]: Done :)

Date: Fri, 3 Jun 2016 10:58:51 -0700

**Comment by redheadedrod** [comment]
Reverted my branch... Closing this for now. Will update my branch with your latest changes and try to resubmit.
